server/src/main/java/org/candlepin/controller/ContentAccessManager.java

```diff
@@ -265,106 +260,126 @@ public ContentAccessCertificate getCertificate(Consumer consumer)
-        // or the payload has changed, regenerate the header.
-        // TODO: we should probably just store this all in one thing rather than two things.
-        if (existing != null) {
-            Date now = new Date();
-            Date expiration = existing.getSerial().getExpiration();
-            Date contentUpdate = owner.getLastContentUpdate();
+        ContentAccessCertificate result;
+        if (existing == null) {
+            result = createNewScaCertificate(consumer, owner);
+        }
+        else {
+            result = updateScaCertificate(consumer, owner, existing);
+        }
 
-            if (expiration.before(now) || !contentUpdate.before(existing.getUpdated()) ||
-                oeca == null || !contentUpdate.before(oeca.getUpdated())) {
+        return result;
+    }
 
-                consumer.setContentAccessCert(null);
-                this.contentAccessCertificateCurator.delete(existing);
+    private ContentAccessCertificate updateScaCertificate(Consumer consumer, Owner owner,
+        ContentAccessCertificate existing) throws GeneralSecurityException, IOException {
+        Date now = new Date();
+        Date expiration = existing.getSerial().getExpiration();
 
-                regenerate = true;
-            }
+        // regenerate x509
+        if (expiration.before(now)) {
+            Validity oneYearValidity = Validity.oneYear();
+            KeyPair keyPair = this.keyPairCurator.getConsumerKeyPair(consumer);
+            CertificateSerial serial = createSerial(oneYearValidity);
+            existing.setSerial(serial);
+            existing.setCert(createX509Cert(consumer, owner, serial, keyPair, oneYearValidity));
+            this.contentAccessCertificateCurator.saveOrUpdate(existing);
+        }
+        // regenerate content
+        Date contentUpdate = owner.getLastContentUpdate();
+        if (!contentUpdate.before(existing.getUpdated())) {
+            existing.setContent(this.createPayloadAndSignature(owner, consumer));
+            this.contentAccessCertificateCurator.saveOrUpdate(existing);
         }
+        return existing;
+    }
 
-        if (existing == null || regenerate) {
-            log.info("Generating new SCA certificate for organization \"{}\"...", owner.getKey());
+    private ContentAccessCertificate createNewScaCertificate(Consumer consumer, Owner owner)
+        throws IOException, GeneralSecurityException {
+        log.info("Generating new SCA certificate for organization \"{}\"...", owner.getKey());
+        ContentAccessCertificate newCert = createScaCert(consumer, owner);
+        ContentAccessCertificate savedCert = this.contentAccessCertificateCurator.create(newCert);
+        consumer.setContentAccessCert(savedCert);
+        this.consumerCurator.merge(consumer);
+        return savedCert;
+    }
+
+    /**
+     * Represents a duration of certificate validity
+     */
+    private static class Validity {
```

Comment:
I feel this class, and the way its used currently, is unnecessarily fancy given the goal here. If we're going to start "fixing" the way we juggle date and times, I'd like to see a proper utility making use of the java.time package which supplants all the date junk we've been doing up to now -- which would be another PR.

Reply:
Sure, moving to time API would help in some aspects. We could even start using it in places like this, but it will not bring many benefits until we start using it in the entities so that we don't have to convert between the two API. The reason for this class is that startDate and endDate need to be created and passed together. I guess it could be modeled as two methods instead but that felt like even more trouble so I picked a value object to encapsulate this date operation with a possibility to parameterize it later.